tasks/main.yml

```diff
@@ -1,39 +1,79 @@
-  include_vars: "{{ __template_vars_file }}"
-  loop:
-    - "{{ ansible_facts['os_family'] }}.yml"
-    - "{{ ansible_facts['distribution'] }}.yml"
-    - "{{ ansible_facts['distribution'] ~ '_' ~
-          ansible_facts['distribution_major_version'] }}.yml"
-    - "{{ ansible_facts['distribution'] ~ '_' ~
-          ansible_facts['distribution_version'] }}.yml"
+- name: Set version specific variables
+  include_vars: "{{ lookup('first_found', ffparams) }}"
   vars:
-    __template_vars_file: "{{ role_path }}/vars/{{ item }}"
-  when: __template_vars_file is file
+    ffparams:
+      files:
+        - "{{ ansible_facts['distribution'] }}_\
+          {{ ansible_facts['distribution_version'] }}.yml"
+        - "{{ ansible_facts['distribution'] }}_\
+          {{ ansible_facts['distribution_major_version'] }}.yml"
+        - "{{ ansible_facts['distribution'] }}.yml"
+        - "{{ ansible_facts['os_family'] }}.yml"
+        - "default.yml"
+      paths:
+        - "{{ role_path }}/vars"
 
-# Examples of some tasks:
 - name: Ensure required packages are installed
   package:
-    name: "{{ __template_packages }}"
+    name: "{{ item }}"
     state: present
+  loop: "{{ __vpn_packages }}"
+
+- name: set the shared_key
+  set_fact:
+    vpn_connections: |
+      {% for connection in vpn_connections %}
+      {%   set _ = connection.__setitem__(
+        "enc_type", connection.enc_type | d(vpn_enc_type)
+        ) %}
+      {%   if connection.enc_type == "psk" %}
+      {%     if "shared_key_content" not in connection %}
+      {%       if "shared_key_src" in connection %}
+      {%         set _ = connection.__setitem__(
+        "shared_key_content", lookup('file', connection.shared_key_src)
+        ) %}
+      {%       endif %}
+      {%     endif %}
+      {%   endif %}
+      {% endfor %}
+      {{ vpn_connections }}
 
-- name: Ensure required services are enabled and started
+- name: show connections
+  debug:
+    msg: conns {{ vpn_connections | to_nice_json }}
+
+- name: build tunnels
+  vars:
+    tunnels: |
+      {% set ns = namespace(unique_tunnels=[]) %}
+      {% for connection in vpn_connections %}
+      {%   if inventory_hostname in connection.hosts %}
+      {%     for node in connection.hosts %}
+      {%       if node != inventory_hostname %}
+      {%         set _ = ns.unique_tunnels.append(node) %}
+      {%       endif %}
+      {%     endfor %}
+      {%   endif %}
+      {% endfor %}
+      {{ ns.unique_tunnels | unique }}
+  block:
+    - name: create ipsec.conf files
+      template:
+        src: "{{ vpn_provider }}.connection.conf"
+        dest: "/etc/ipsec.d/{{ inventory_hostname }}-to-{{ item }}.conf"
+        mode: '0644'
+      with_items: "{{ tunnels }}"
+
+    - name: create ipsec.secrets files
+      template:
+        src: "{{ vpn_provider }}.connection.secrets"
+        dest: "/etc/ipsec.d/{{ inventory_hostname }}-to-{{ item }}.secrets"
+        mode: '0600'
+      with_items: "{{ tunnels }}"
```

Comment:
The next step will be to use handlers here - the `template` tasks will have a `notify: __vpn_handler_restart_services`
in the `handlers/` directory you'll have a file called `main.yml` with tasks like this:
```yaml
- name: enable and restart vpn services
  listen: __vpn_handler_restart_services
  service: ....
```
That way the role will be idempotent - if you run the role again with the same parameters, the `template` tasks should report no changes, which means the handler won't be called, which means you won't be unnecessarily restarting the vpn services.